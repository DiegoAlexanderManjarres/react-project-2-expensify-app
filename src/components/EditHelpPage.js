import React from 'react';

const EditHelpPage = () => (
  <div>My edit help page</div>
)

export default EditHelpPage;