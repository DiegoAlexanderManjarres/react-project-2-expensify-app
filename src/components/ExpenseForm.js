import React from 'react'
import Parser from 'html-react-parser'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
 	constructor(props) {
		super(props)
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: '',
			titleButton: props.expense? 'Save Expense' : 'Add Expense'
		}
  }
  
   onDescriptionChange = (e) => {
		e.preventDefault();
		const description = e.target.value;
		this.setState(() => ({ description }));
  }
  	onNoteChange = (e) => {
		e.preventDefault()  
		const note = e.target.value;
		this.setState(() => ({ note }))
  }
 	onAmountChange = (e) => {
		e.preventDefault();
		const reg = /^\d{1,}(\.\d{0,2})?$/;
		const amount = e.target.value;
		if (!amount || amount.match(reg)) {
			this.setState(() => ({ amount }));
		} 
  }
  onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}    
   }
  	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }))
   }
   onSubmit = (e) => {
		e.preventDefault();
		if (!this.state.description || !this.state.amount) {
		// Set error state to 'Please provide description and amount'
		// used module html-react-parser for the template string with html
			this.setState(() => ({
				...this.state, 
				error: `
				   Please provide 
					<strong> description </strong> 
					and 
					<strong> amount </strong>.
				`
			}));
		} else {
		// Clear the error
			this.setState(() => ({ ...this.state, error: '' }));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			})
		}
   }
   render(){
		return (	
			<form className="form" onSubmit={this.onSubmit}>
				{
					this.state.error && 
					<p className="form__error">{Parser(this.state.error)}</p>
				}
					<input
						type='text'
						placeholder='Description'
						autoFocus
						className="text-input"
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input 
						type="text"
						placeholder="Amount"
						className="text-input"
						value={this.state.amount}
						onChange={this.onAmountChange}            
					/>
					<SingleDatePicker 
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
						block
						/>
					<textarea 
						placeholder="Add a note for your expense (optional)"
						value={this.state.note}
						onChange={this.onNoteChange}
						className="textarea"/>
					<div>
						<button className="button">{this.state.titleButton}</button>
					</div>	
			</form>
			
		)
  }
}
