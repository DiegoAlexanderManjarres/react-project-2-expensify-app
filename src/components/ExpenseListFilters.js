import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize'
import uuidv4 from 'uuidv4'
import { 
	setTextFilter, 
	sortByAmount, 
	sortByDate, 
	setStartDate,
	setEndDate 
} from '../actions/filters'


export class  ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null,
		idGennerator: uuidv4()
	}  
	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate)
		this.props.setEndDate(endDate)
	}
	onFocusChange = (calendarFocused) => (
		this.setState(() => ({ calendarFocused }))
	)
	onTextChange = (e) => {
		const text = e.target.value;
		this.props.setTextFilter(text);
	}
	onSortChange = (e) => (
		e.target.value === 'date' 
		? this.props.sortByDate() 
		: this.props.sortByAmount()        
	)		
	render() {
		return (
			<div className="content-container">
				<div className="input-group">
					<div className="input-group__item">
						<input 
							type='text' 
							value={this.props.filters.text}
							onChange={this.onTextChange}
							className="select"
							placeholder="Search Expenses"/>							
					</div>
					<div className="input-group__item">
						<select
							onChange={this.onSortChange}
							className="text-input">
							<option value="date">Date</option>
							<option value="amount">Amount</option>
						</select>
					</div>
					<div className="input-group__item">
						<DateRangePicker 
							startDate={this.props.filters.startDate}
							endDate={this.props.filters.endDate}
							onDatesChange={this.onDatesChange}
							focusedInput={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							numberOfMonths={1}
							isOutsideRange={() => false}
							showClearDates={true}
							startDateId={this.state.idGennerator}
							endDateId={this.state.idGennerator}
						/>
					</div>
				</div>
			</div>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: (amount) => dispatch(sortByAmount(amount)),
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

const mapStateToProps = (state) => ({
	filters: state.filters
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)