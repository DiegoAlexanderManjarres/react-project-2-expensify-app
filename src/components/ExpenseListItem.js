// Export a Stateless functional component
// description, amount, createdAt
import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import moment from 'moment'
import numeral from 'numeral'

export const ExpenseListItem = ({id, description, amount, createdAt}) => {
   return (    
	  	<li  data-id={id}>
			<h3><Link to={`/edit/${id}`}>{description}</Link></h3>
			<p>
				<strong> Amount: </strong> 
					{numeral(amount / 100).format('$0,0.00')} 
				<strong> CreatedAt: </strong> 
					{moment(createdAt).format('MMMM Do, YYYY')}
			</p>        
	  	</li>   
   )
}
export default connect()(ExpenseListItem);