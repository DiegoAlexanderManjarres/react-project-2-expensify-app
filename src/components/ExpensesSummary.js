import React from 'react'
import { connect } from 'react-redux'
import selectExpenseTotal from '../selectors/Exenses-total'
import selectExpenses from '../selectors/expenses'
import numeral from 'numeral'

export const ExpensesSummary = ({ expenses }) => {
   const total = numeral(selectExpenseTotal(expenses) / 100).format('$0,0.00')
   return (            
      <React.Fragment>
         { expenses.length ?
            <div>
               <div>Summary</div>
               <p><strong> Expenses available: </strong>{expenses.length}</p>
               <p><strong> Expenses Total: </strong>{total}</p>
            </div>
            : <div>Add expense</div>
         }     
      </React.Fragment>       
   )
} 

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpensesSummary);
