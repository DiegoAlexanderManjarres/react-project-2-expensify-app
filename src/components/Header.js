import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <div>
      <NavLink to="/" activeClassName="is-active" exact={true}>
        <strong> Dashboard </strong> 
      </NavLink>
      <br />
      <NavLink to="/create" activeClassName="is-active"> 
        <strong> Create-Expense </strong> 
      </NavLink>
    </div>
  </header>
)

export default Header;