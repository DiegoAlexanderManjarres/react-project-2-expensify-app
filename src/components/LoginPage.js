import React from 'react'
import { connect } from 'react-redux'
import { startLogin } from '../actions/auth'


export const LoginPage = ({ startLogin }) =>  (
		<section className="box-layout">
			<div className="box-loyout__box">
				<h1 className="box-layout__title">Esxpensify</h1>
				<p>Its time to take charge and track your expenses.</p>
				<button  
					className="button" 
					onClick={startLogin}
					>Login with Google
				</button>
			</div>			
		</section>
	)


const mapDispatchToProps = (dispatch) => ({
	startLogin: () => dispatch(startLogin())
})   
export default connect(undefined, mapDispatchToProps)(LoginPage)   