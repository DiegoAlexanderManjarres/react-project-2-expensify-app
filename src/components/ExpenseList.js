import React from "react";
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
// connect for components that needs to read from the store or didpatch actions

export const ExpenseList = (props) => (
	<div>
		{
			!props.expenses.length ? (<p>No expenses</p>) : (
				<ul>    
					{
						props.expenses.map(expense => (
							<ExpenseListItem 
								key={expense.id} 
								{...expense} 
								{...props}/>
						))
					}
				</ul>    
			)
		}
			
	</div>
);

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);