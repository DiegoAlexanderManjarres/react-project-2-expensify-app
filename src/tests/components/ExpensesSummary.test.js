import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'


test('Viewing 2 expenses to talling $94.34', () => {
   const expenses = [{ amount: 2000 }, { amount: 7434 }]
   const wrapper = shallow(<ExpensesSummary expenses={expenses}/>)
   expect(wrapper).toMatchSnapshot()
})

test('Viewing 1 expenses to talling $94.34', () => {
   const expenses = [{ amount: 9434 }]
   const wrapper = shallow(<ExpensesSummary expenses={expenses}/>)
   expect(wrapper).toMatchSnapshot()
})

test('Viewing 0 expenses', () => {
   const expenses = []
   const wrapper = shallow(<ExpensesSummary expenses={expenses}/>)
   expect(wrapper).toMatchSnapshot()
})



