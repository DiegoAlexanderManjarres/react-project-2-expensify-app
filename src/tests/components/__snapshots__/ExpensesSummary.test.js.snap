// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewing 0 expenses 1`] = `
<Fragment>
  <div>
    Add expense
  </div>
</Fragment>
`;

exports[`Viewing 1 expenses to talling $94.34 1`] = `
<Fragment>
  <div>
    <div>
      Summary
    </div>
    <p>
      <strong>
         Expenses available: 
      </strong>
      1
    </p>
    <p>
      <strong>
         Expenses Total: 
      </strong>
      $94.34
    </p>
  </div>
</Fragment>
`;

exports[`Viewing 2 expenses to talling $94.34 1`] = `
<Fragment>
  <div>
    <div>
      Summary
    </div>
    <p>
      <strong>
         Expenses available: 
      </strong>
      2
    </p>
    <p>
      <strong>
         Expenses Total: 
      </strong>
      $94.34
    </p>
  </div>
</Fragment>
`;
